dbms/src/Functions/FunctionsJson.h

```diff
@@ -530,11 +612,671 @@ class FunctionsJsonArray : public IFunction
+        }
+
+        block.getByPosition(result).column = std::move(col_to);
+    }
+
+private:
+    template <typename FromDataType, bool is_timestamp>
+    static void doExecute(
+        ColumnString::Chars_t & data_to,
+        ColumnString::Offsets & offsets_to,
+        const ColumnPtr & column_ptr_from)
+    {
+        const auto * column_from
+            = checkAndGetColumn<ColumnVector<typename FromDataType::FieldType>>(column_ptr_from.get());
+        RUNTIME_CHECK(column_from);
+        const auto & data_from = column_from->getData();
+        size_t reserve_size = data_from.size() * (1 + 1 + sizeof(UInt64));
+        JsonBinary::JsonBinaryWriteBuffer write_buffer(data_to, reserve_size);
+        for (size_t i = 0; i < data_from.size(); ++i)
+        {
+            if constexpr (std::is_same_v<DataTypeMyDate, FromDataType>)
+            {
+                MyDate date(data_from[i]);
+                JsonBinary::appendDate(write_buffer, date);
+            }
+            else
+            {
+                MyDateTime date_time(data_from[i]);
+                if constexpr (is_timestamp)
+                    JsonBinary::appendTimestamp(write_buffer, date_time);
+                else
+                    JsonBinary::appendDatetime(write_buffer, date_time);
+            }
+
+            writeChar(0, write_buffer);
+            offsets_to[i] = write_buffer.count();
+        }
+    }
+
+private:
+    const tipb::FieldType * input_tidb_tp = nullptr;
+};
+
+class FunctionCastDurationAsJson : public IFunction
+{
+public:
+    static constexpr auto name = "cast_duration_as_json";
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionCastDurationAsJson>(); }
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool useDefaultImplementationForNulls() const override { return true; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if unlikely (!arguments[0]->isMyTime())
+            throw Exception(
+                fmt::format("Illegal type {} of argument of function {}", arguments[0]->getName(), getName()),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        return std::make_shared<DataTypeString>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        auto col_to = ColumnString::create();
+        auto & data_to = col_to->getChars();
+        auto & offsets_to = col_to->getOffsets();
+        auto rows = block.rows();
+        offsets_to.resize(rows);
+
+        const auto & from = block.getByPosition(arguments[0]);
+        if (const auto * duration_type = checkAndGetDataType<DataTypeMyDuration>(from.type.get());
+            likely(duration_type))
+        {
+            auto fsp = duration_type->getFsp();
```

Comment:
Looks like in TiDB's impl, the fsp is always set to MaxFSP
https://github.com/pingcap/tidb/blob/3543275dcf4b6454eb874c1362c87d31a963da6d/pkg/expression/builtin_cast.go#L921

Reply:
ok, updated.